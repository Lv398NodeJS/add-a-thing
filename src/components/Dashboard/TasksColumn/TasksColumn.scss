.tasks-column-todo-title,
.tasks-column-in-progress-title,
.tasks-column-done-title,
.tasks-column-invalid-title {
  text-align: center;
  font-size: 1.8rem;
  user-select: none;
}

.tasks-column-todo-title {
  color: rgb(103, 102, 106);
  border-bottom: 3px solid rgb(103, 102, 106);
}

.tasks-column-in-progress-title {
  color: rgb(118, 183, 192);
  border-bottom: 3px solid rgb(116, 154, 159);
}

.tasks-column-done-title {
  color: rgb(194, 105, 95);
  border-bottom: 3px solid rgb(191, 134, 123);
}

.tasks-column-invalid-title {
  color: rgb(78, 26, 235);
  border-bottom: 3px solid rgb(78, 26, 235);
}

.task-items-container {
  flex: 0 1 auto;
  overflow-y: auto;
  padding: 0px;
  width: 100%;
  max-height: 460px;
}

::-webkit-scrollbar {
  display: none;
}
