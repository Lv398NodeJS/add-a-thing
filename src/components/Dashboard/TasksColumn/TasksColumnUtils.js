export const columnTitleClass = (title) => {
  switch (title) {
    case 'To Do':
      return 'tasks-column-todo-title';
    case 'In Progress':
      return 'tasks-column-in-progress-title';
    case 'Done':
      return 'tasks-column-done-title';
    default:
      return 'tasks-column-invalid-title';
  }
};

export const loaderColor = (title) => {
  switch (title) {
    case 'To Do':
      return '#67666a';
    case 'In Progress':
      return '#749a9f';
    case 'Done':
      return '#bf867b';
    default:
      return '#4e1aeb';
  }
};
