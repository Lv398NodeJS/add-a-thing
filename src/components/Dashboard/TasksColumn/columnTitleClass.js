export default (title) => {
  switch (title) {
    case 'To Do':
      return 'tasksColumnToDoTitle';
    case 'In Progress':
      return 'tasksColumnInProgressTitle';
    case 'Done':
      return 'tasksColumnDoneTitle';
    default:
      return 'tasksColumnInvalidTitle';
  }
};
