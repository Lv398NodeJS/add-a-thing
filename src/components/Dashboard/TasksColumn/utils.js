export const columnTitleClass = (title) => {
  switch (title) {
    case 'To Do':
      return 'tasks-column-todo-title';
    case 'In Progress':
      return 'tasks-column-in-progress-title';
    case 'Done':
      return 'tasks-column-done-title';
    default:
      return 'tasks-column-invalid-title';
  }
};
