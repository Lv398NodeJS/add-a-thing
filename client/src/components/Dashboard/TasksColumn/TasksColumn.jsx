import React, { Component } from 'react';
import { Container } from 'react-bootstrap';
import { connect } from 'react-redux';
import Loader from 'react-loader-spinner';
import { columnTitleClass, loaderColor } from './TasksColumnUtils';
import TaskItem from '../TaskItem/TaskItem';
import SortList from '../SortList/SortList';
import { sortComparer } from '../SortList/sortUtils';
import './TasksColumn.scss';

class TasksColumn extends Component {
  onDrop = (event) => {
    event.preventDefault();

    const { handleTaskDrop } = this.props;

    const taskID = event.dataTransfer.getData('taskID');
    const newStatus = event.currentTarget.dataset.status;

    handleTaskDrop(taskID, newStatus);

    const fakeTask = document.getElementById('drag-avatar');
    if (fakeTask != null) fakeTask.remove();
    document.getElementById('delete-zone').classList.remove('shown');
  };

  render() {
    const {
      title, filteredTasks, loading, allSortSettings = {},
    } = this.props;

    const sortIconColor = loaderColor(title);

    const storageKey = `${document.URL.split('/').pop()}:${title}`;
    const sortState = allSortSettings[storageKey] || {};
    const sortedTasks = [...filteredTasks].sort(sortComparer(sortState.field, sortState.direction));

    const tasksToDisplay = sortedTasks.map(
      task => (
        <TaskItem
          key={task._id}
          status={task.status}
          priority={task.priority}
          id={task._id}
          taskName={task.name}
        />
      ),
    );

    const loader = (
      <Loader
        type="ThreeDots"
        color={loaderColor(title)}
        height="80"
        width="100"
        className="mt-5"
      />
    );

    return (
      <div className="tasks-column rounded mb-4 mb-lg-0">
        <h1
          data-test="columnTitle"
          className={columnTitleClass(title)}
        >
          {title}
          <SortList
            storageKey={storageKey}
            color={sortIconColor}
          />
        </h1>
        <Container
          fluid="true"
          onDragOver={e => e.preventDefault()}
          onDrop={this.onDrop}
          data-status={title}
          status={title}
          className="task-items-container h-100 px-4 pb-4"
          data-test="taskItemsContainer"
        >
          {loading ? loader : tasksToDisplay}
        </Container>
      </div>
    );
  }
}

const mapStateToProps = ({ mainContainerReducer: { taskListRef, loading }, sortListReducer }) => ({
  taskListRef,
  loading,
  allSortSettings: sortListReducer,
});

export { TasksColumn as TaskColumnComponent };
export default connect(mapStateToProps)(TasksColumn);
