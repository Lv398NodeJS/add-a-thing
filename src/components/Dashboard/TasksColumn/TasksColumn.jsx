import React, { Component } from 'react';
import { Container } from 'react-bootstrap';
import './TasksColumn.scss';
import TaskItem from '../TaskItem/TaskItem';
import SortList from '../SortList/SortList';

export default class TasksColumn extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  columnTitleClass = () => {
    const { title } = this.props;

    switch (title) {
      case 'To Do':
        return 'tasksColumnToDoTitle';
      case 'In Progress':
        return 'tasksColumnInProgressTitle';
      case 'Done':
        return 'tasksColumnDoneTitle';
      default:
        return 'tasksColumnInvalidTitle';
    }
  };

  sortIconColor = () => {
    const { title } = this.props;

    switch (title) {
      case 'To Do':
        return 'rgb(103, 102, 106)';
      case 'In Progress':
        return 'rgb(118, 183, 192)';
      case 'Done':
        return 'rgb(194, 105, 95)';
      default:
        return 'rgb(78, 26, 235)';
    }
  };

  render() {
    const {
      sortedTasks, taskListRef, title,
    } = this.props;

    const sortIconColor = this.sortIconColor();

    const tasksToDisplay = sortedTasks.map(
      task => (
        <TaskItem
          key={task.id}
          status={task.status}
          priority={task.priority}
          priorityForSorting={['Low', 'Medium', 'High'].indexOf(task.priority)}
          id={task.id}
          taskName={task.name}
          taskListRef={taskListRef}
        />
      ),
    );

    return (
      <div className="tasksColumn rounded mb-3">
        <h1 className={this.columnTitleClass()}>
          {title}
        </h1>
        <Container className="taskItemsContainer">
          <SortList
            storageKey={window.location.pathname + title}
            sortIconColor={sortIconColor}
            fields={[
              {
                key: 'taskName',
                text: 'Name',
              },
              {
                key: 'priorityForSorting',
                text: 'Priority',
              },
            ]}
          >
            {tasksToDisplay.reverse()}
          </SortList>
        </Container>
      </div>
    );
  }
}
