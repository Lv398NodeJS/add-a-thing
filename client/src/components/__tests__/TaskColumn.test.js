import React from 'react';
import { shallow } from 'enzyme';
import * as utils from '../Dashboard/TasksColumn/TasksColumnUtils';
import { TaskColumnComponent } from '../Dashboard/TasksColumn/TasksColumn';

const taskColumn = shallow(<TaskColumnComponent
  filteredTasks={[]}
  taskListRef=""
  title="test"
/>);

describe('<TaskColumn />', () => {
  it('should render and match snapshot', () => {
    const taskColumnR = shallow(<TaskColumnComponent filteredTasks={[]} />);
    expect(taskColumnR).toMatchSnapshot();
  });
  it('should render title element', () => {
    const title = taskColumn.find('[data-test="columnTitle"]');
    expect(title.length).toEqual(1);
  });
  it('should render title text from props', () => {
    const titleText = 'To Do';
    const taskColumnT = shallow(<TaskColumnComponent filteredTasks={[]} title={titleText} />);
    const title = taskColumnT.find('[data-test="columnTitle"]').text();
    expect(title).toEqual(titleText);
  });
  it('should have correct ToDo class name', () => {
    const titleText = 'To Do';
    const taskColumnT = shallow(<TaskColumnComponent filteredTasks={[]} title={titleText} />);
    const title = taskColumnT.find('[data-test="columnTitle"]');
    const finalTitleStyle = utils.columnTitleClass(titleText);
    expect(title.hasClass(finalTitleStyle)).toBeTruthy();
  });
  it('should have correct InProgress class name', () => {
    const titleText = 'In Progress';
    const taskColumnT = shallow(<TaskColumnComponent filteredTasks={[]} title={titleText} />);
    const title = taskColumnT.find('[data-test="columnTitle"]');
    const finalTitleStyle = utils.columnTitleClass(titleText);
    expect(title.hasClass(finalTitleStyle)).toBeTruthy();
  });
  it('should have correct Done class name', () => {
    const titleText = 'Done';
    const taskColumnT = shallow(<TaskColumnComponent filteredTasks={[]} title={titleText} />);
    const title = taskColumnT.find('[data-test="columnTitle"]');
    const finalTitleStyle = utils.columnTitleClass(titleText);
    expect(title.hasClass(finalTitleStyle)).toBeTruthy();
  });
  it('should render container for tasks', () => {
    const tasksCont = taskColumn.find('[data-test="taskItemsContainer"]');
    expect(tasksCont.length).toBe(1);
  });
});
