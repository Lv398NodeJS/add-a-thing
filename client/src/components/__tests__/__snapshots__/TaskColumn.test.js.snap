// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskColumn /> should render and match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TasksColumn
    filteredTasks={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1
          className="tasks-column-invalid-title"
          data-test="columnTitle"
        >
          <UNDEFINED
            color="#4e1aeb"
            storageKey=":undefined"
          />
        </h1>,
        <ForwardRef
          className="task-items-container h-100 px-4 pb-4"
          data-test="taskItemsContainer"
          fluid="true"
          onDragOver={[Function]}
          onDrop={[Function]}
        />,
      ],
      "className": "tasks-column rounded mb-4 mb-lg-0",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            undefined,
            <UNDEFINED
              color="#4e1aeb"
              storageKey=":undefined"
            />,
          ],
          "className": "tasks-column-invalid-title",
          "data-test": "columnTitle",
        },
        "ref": null,
        "rendered": Array [
          undefined,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "color": "#4e1aeb",
              "storageKey": ":undefined",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(SortList)",
              "type": [Function],
              Symbol(Symbol.hasInstance): [Function],
            },
          },
        ],
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [],
          "className": "task-items-container h-100 px-4 pb-4",
          "data-status": undefined,
          "data-test": "taskItemsContainer",
          "fluid": "true",
          "onDragOver": [Function],
          "onDrop": [Function],
          "status": undefined,
        },
        "ref": null,
        "rendered": Array [],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "fluid": false,
          },
          "displayName": "Container",
          "render": [Function],
        },
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1
            className="tasks-column-invalid-title"
            data-test="columnTitle"
          >
            <UNDEFINED
              color="#4e1aeb"
              storageKey=":undefined"
            />
          </h1>,
          <ForwardRef
            className="task-items-container h-100 px-4 pb-4"
            data-test="taskItemsContainer"
            fluid="true"
            onDragOver={[Function]}
            onDrop={[Function]}
          />,
        ],
        "className": "tasks-column rounded mb-4 mb-lg-0",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              undefined,
              <UNDEFINED
                color="#4e1aeb"
                storageKey=":undefined"
              />,
            ],
            "className": "tasks-column-invalid-title",
            "data-test": "columnTitle",
          },
          "ref": null,
          "rendered": Array [
            undefined,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "color": "#4e1aeb",
                "storageKey": ":undefined",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.memo),
                "WrappedComponent": [Function],
                "compare": null,
                "displayName": "Connect(SortList)",
                "type": [Function],
                Symbol(Symbol.hasInstance): [Function],
              },
            },
          ],
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [],
            "className": "task-items-container h-100 px-4 pb-4",
            "data-status": undefined,
            "data-test": "taskItemsContainer",
            "fluid": "true",
            "onDragOver": [Function],
            "onDrop": [Function],
            "status": undefined,
          },
          "ref": null,
          "rendered": Array [],
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "fluid": false,
            },
            "displayName": "Container",
            "render": [Function],
          },
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
